src/libsystemd/sd-device/sd-device.c

```diff
@@ -183,6 +183,33 @@ int device_set_syspath(sd_device *device, const char *_syspath, bool verify) {
                         return r;
                 }
 
+                if (!path_startswith(syspath, "/sys")) {
+                        _cleanup_free_ char *real_sys = NULL, *relpath = NULL;
+
+                        /* /sys is a symlink to somewhere sysfs is mounted on? In that case, we convert the path to real sysfs to "/sys". */
+                        if (!is_symlink("/sys"))
+                                return -ENODEV;
+
+                        r = readlink_and_canonicalize("/sys", NULL, &real_sys);
+                        if (r < 0)
+                                return log_debug_errno(r, "sd-device: could not read symlink /sys: %m");
+
+                        if (!path_startswith(syspath, real_sys)) {
+                                log_debug("sd-device: canonicalized path '%s' does not starts with '%s'. Something strange...", syspath, real_sys);
+                                return -ENODEV;
+                        }
+
+                        r = path_make_relative(real_sys, syspath, &relpath);
```

Comment:
as above, i'd just use chase_symlinks directly on /sys, and then chop off that from the new path here. i.e. now manual invocation of readlink_and_canonicalize or so so, but just always use chase_symlinks(), to determine the real meaning of both the specified syspath and of /sys